### ISSUE TITLE
[CURATOR-14] Memory leak in Curator watches

### ISSUE THREAD

**Issue by jira-importer** [apache/curator#535]
<p>The JVM runs out of memory if you repetitively create a PathChildrenCache, start it then immediately stop it.  It appears that the memory is taken up by a watch that isn't ever cleaned up.  Curator attempts to do some pooling of watches, but doesn't seem to use the path in the pooling.</p>

---
<details><summary><i>Originally reported by <a title="Brandon Beck" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bbeck">bbeck</a>, imported from: <a href="https://issues.apache.org/jira/browse/CURATOR-14" target="_blank">Memory leak in Curator watches</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Ioannis Canellos" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel">iocanel</a>
<li><b>status</b>: Open
<li><b>priority</b>: Minor
<li><b>resolution</b>: Unresolved
<li><b>imported</b>: 2025-01-21
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bbeck">bbeck</a>:</i>
<p>With this test I can reliable run out of memory (8mb heap) in a minute or so.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>This will be difficult to fix unless Curator adds some support for pooled watches. AFAIK, there's no way to clear a watch in ZooKeeper. So, it could be argued, that this is a client problem. You could get the same results by using ZooKeeper directly. i.e. Add new watches in a loop.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>The only way to "fix" this is to support a Watcher pool of some kind. It wouldn't be hard to add. Just a lot of search/replace work. </p>

<p>Off the top of my head - I already have the beginnings of a Watcher cache in the private class NamespaceWatcherMap. This could be extended so that it's exposed via the CuratorFramework instance. Recipes would use the new cache API to allocate watchers instead of manually allocating them.</p>

<p>hint - hint - whomever implements this would make a very good candidate to be a Curator committer.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel">iocanel</a>:</i>
<p>I tried the suggested approach about caching watchers and it seems to work. The only problem is providing a working equals() and hashCode() for the Watchers implemented as anonymous inner classes. So I think that something like a CacheableWatcher that could provide an impl of these two based on a key property could reduce the boilerplate.</p>

<p>If no one is already working on it I could provide a patch.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel">iocanel</a>:</i>
<p>A first draft.</p>

<p>The patch adds the WatcherCache interface with a default implementation the NamespaceWatcherCache (used to be NamespaceWatcherMap). The cache is expose by the client as Jason described.</p>

<p>Recipes have been tuned to use the WatcherCache where it makes sense.</p>

<p>Tested against the attached case and it solves the memory leak.</p>

<p>This implementation of course still has issues when multiple instances of the same recipe are used. Need to work on that more.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel">iocanel</a>:</i>
<p>I am attaching a second draft of the patch.</p>

<p>There are still things need to be done, but I'd like some feedback/guidelines.</p>

<p>The patch modifies the NamespaceWatcherMap and adds it some pooling characteristics (a borrow and release method backed by an array list).</p>

<p>The main issue I have is when to "borrow" and when to "release":</p>

<p>i) Having the recipes do the "borrow" and "release" in some cases is not enough as not always is obvious when to "release". <br/>
ii) Releasing when the watcher processes the event again is not enough, as I think that curator may reuse Watcher instance again in some cases (not 100% sure).<br/>
iii) "borrow" inside the Watching getWatcher() seems like a good idea but then again its not easy to "release".</p>

<p>Thoughts?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel">iocanel</a>:</i>
<p>This is the 3rd draft and hopefully final<img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/help_16.png" height="16" width="16" align="absmiddle" alt="" border="0"/> version of the patch.</p>

<p>The patch adds a WatcherPool for pooling NamespaceWatchers. The NamespaceWatcherMap is being backed by the WatcherPool: It pulls NamespaceWatcher instances from the pool when creates new mappings.</p>

<p>Upon removal of a mapping the NamespaceWatcher (if not garbage collected is returned to the pool).</p>

<p>Recipes allocate NamespaceWatchers from the map and remove them from the map when no longer needed.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>comments:</p>

<ul>
	<li>WatcherMap should extend Closeable</li>
	<li>JavaDoc @return on WatcherMap is empty</li>
	<li>Why is WatcherMap.get() parameterized? Shouldn't it just return Watcher?</li>
	<li>NamespaceWatcher.warp() is misspelled</li>
	<li>DistributedDoubleBarrier.acutalWatcher is misspelled</li>
	<li>I think that WatcherMap should introduce a new Interface other than re-using Watcher. It's a bit confusing to use a ZooKeeper Watcher as a key. Maybe a new interface called WatcherKey or something. I realize that this is how I had it but now that it's getting distributed throughout the code base I can see that it's confusing. This will also prevent a bad watcher from getting passed to Curator/ZooKeeper.</li>
	<li>Why is LockInternals.internalWatcherMap needed?</li>
</ul>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel">iocanel</a>:</i>
<p>I had 2 kind of issues on the LockInternals:<br/>
i) I was unsure on when I should remove the watcher from the global map.<br/>
ii) I was unsure if its possibly that a watcher for the same path can be reused.</p>

<p>For <img class="emoticon" src="https://issues.apache.org/jira/images/icons/emoticons/information.png" height="16" width="16" align="absmiddle" alt="" border="0"/> I thought that the most fitting place would be to remove the watcher, when it fired (and maybe when we release the lock?).<br/>
For (ii) I used the internal map. I am totally unsure about that.</p>

<p>So please share your thoughts/ideas about it.</p>

<p>Regarding the WatcherKey, do u see it as a way of wrapping Watcher/CuratorWatcher objects or you think that the recipes should directly use anonymous inner implementations of the WatcherKey?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>>Regarding the WatcherKey, do u see it as a way of wrapping Watcher/CuratorWatcher objects or you think that the recipes should directly use anonymous inner implementations of the WatcherKey?<br/>
Yes. The idea is to have a semi-opaque type that isn't accepted in any Curator/ZooKeeper API. This way, the user can't make a mistake. What do you think?</p>

<p>Thinking about this more - I wonder if we could hide the WatcherMap altogether inside of Curator. Extend with usingWatcher() semantics to take a WatcherKey instance. This would be translated internally into a WatcherMapped instance. I haven't thought this through. What do you think?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>In my implementation of NamespaceWatcherMap there isn't a need to return the watcher. It uses a Guava cache with weak values. Do you see a need to return watchers to the pool?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel">iocanel</a>:</i>
<p>If we don't use a Pool of NamespaceWatcher how do we solve the problem of reusing NamespaceWatchers? Isn't that the goal?</p>

<p>From what I understand we can't reuse NamespaceWatcher instances that were mapped to annonymous inner Wathcer/CuratorWatcher if we create a new instance of the recipe, right?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>The pool is done via a Guava cache with weak values. So, the items in the cache automatically expire. There's no need to have borrow/return semantics.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel">iocanel</a>:</i>
<p>Apparently there is something I am missing here :-D</p>

<p>So the main problem is that there is no way to clear a Watcher that has been assigned to ZooKeeper. <br/>
Eventually this can lead to a memory leak since its not possible to garbage collect Watcher instances.<br/>
In order to workaround this problem we should limit the amount of Watchers we create and reuse Watcher instances when its possible.</p>

<p>Currently curator is using Guava's cache that uses "weak values" so that values can be garbage collected (once they are weakly reachable). This is done transparently whenever we call usingWatcher via Watching.</p>

<p>There are a couple of things that I don't understand here:</p>

<p>i) If using "guava's cache" was enough, then what is to be implemented for this issue? The cache is already in place and is already being used in a transparent way.<br/>
ii) Does the garbage collector ever manage to collect values of the cache, given that they are not cleared in ZooKeeper? From some tests I run using a modified version of the attached MemoryTest I don't see any value getting garbage collected.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cheddar">cheddar</a>:</i>
<p>I haven't been too privy to the full context of this, so forgive me if this was discussed somewhere and shot down already.</p>

<p>But, instead of pooling watches, I believe that if there was one large global watcher that could essentially be re-used for everything, then that would significantly reduce the memory footprint required by the ZooKeeper object to store watches.  That is, its internal data structures would just be a bunch of references to the same actual object, so you just pay the cost of the extra references.  Given that it stores things as a Set<> on a path, it would only store it once for each path as well.  This global watcher could have Curator-specific registration mechanisms that allow it to have watchers registered and unregistered so that you can re-create the current behavior and get the semantics that we want.</p>

<p>Does that make sense?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel">iocanel</a>:</i>
<p>Interesting.</p>

<p>Ideally we could use a multimap inside the global watcher for map paths back to the set of registered watchers. If we could also make sure that the watchers registered internally can be garbage collected then we've solved the problem.</p>

<p>I'll experiment with it and see how far I can get.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>FYI - when I first designed Curator I'd planned not to expose Watchers but have everything flow through CuratorListener. This mechanism might be used for this.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>I like this approach. I'm reviewing it now.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>In DispatchingWatcher, why is the key to the watchers a UUID instead of the ZNode path?</p>

<p>More - why can't ids and watchers maps be combined?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel">iocanel</a>:</i>
<p>A specific path can have more than one Watchers, so the path doesn't make a good key.</p>

<p>Ideally, the maps could be combined in one MultiMap<String, Watcher> with soft values, but I couldn't find such implementation, So I used a Multimap<String,String> and a Cache<String, Watcher> with soft values and used a UUID for correlation. I preferred Cache instead of a second Map so that I can use a RemovalListener that would remove entries from the Multimap when they are garbage collected from the Cache.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cheddar">cheddar</a>:</i>
<p>If we are introducing a mechanism that allows for unregistering a listener, I wonder if we need the Soft cache.  A part of me thinks that the semantics should be that recipes are in charge of unregistering their stuff and if they don't, they have a bug.  If we were to take that tact, then it would seem like a Multimap would be just fine, I think?</p>

<p>Or, is there another reason to have the soft values that I'm not thinking of?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel">iocanel</a>:</i>
<p>The latest patch doesn't include a mechanism with unregistering a watcher. It unregisters Watchers that are fired and allows garbage collecting Watchers that are weakly referenced.</p>

<p>I don't have any objection in removing the weak values and adding some logic in the recipes (something like combining the last patch with draft 3).</p>

<p> I am not very confident though on when we should unregister inside:<br/>
i) LockInternals<br/>
ii) Barriers</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>Yeah, there's no way to remove a watcher...</p>

<p>I have an alternate implementation that removes the syncs. As I implemented it I realized that Soft Refs aren't really needed. Watchers stay active until they fire and Ioannis' code already removes the mapped watcher when it fires. Please correct me if I'm wrong.</p>

<p><a href="https://gist.github.com/Randgalt/5988052" class="external-link" target="_blank" rel="nofollow noopener">https://gist.github.com/Randgalt/5988052</a></p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cheddar">cheddar</a>:</i>
<p>I finally had a chance to look at code and not just talk in high level abstractiness, but I don't see why we cannot have an unregister method now (and I wonder if, without that, we aren't just re-implementing the watcher mechanism that ZK already has).</p>

<p>It seems like we should be able to adjust DispatchingWatcher to have an unregister() method on it.  We can then put methods for registering and returning the watcher directly on the CuratorFramework along with methods to unregister the watcher.  Simple object equality should be sufficient for removing the watcher from the Set in DispatchingWatcher.  The big concern here is race conditions around a watch getting fired and it being unregistered concurrently.  I think this is a fine concern to bubble up to the users, they have to write their Watchers/close methods with the understanding that even if they are currently unregistering the watcher, it could still get called.</p>

<p>Then all of the recipes that end up calling the builders with a "usingWatcher" will have the ability to remove their watchers from the base CuratorFramework when their respective close() methods are called.</p>

<p>Hopefully that makes sense?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cheddar">cheddar</a>:</i>
<p>For LockInternals/Barriers, I haven't looked too closely, but perhaps they should have a close() method?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p>I think unregister in combination with DispatchingWatcher would work. The issue that's being solved, BTW, is that there is a kind of memory leak if multiple watchers for the same path keep getting set as there's no way in ZK to remove a watcher. DispatchingWatcher solves this by always using the same Watcher instance.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel">iocanel</a>:</i>
<p>I think that there is consensus in adding & using an unregister mechanism.<br/>
I can work on that once I get back from vacation (2 weeks).</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel" class="user-hover" rel="iocanel">Ioannis Canellos</a> - what do you think of my change that removes syncs?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel">iocanel</a>:</i>
<p>I like it. I think that since we are going to use a single data structure its safe to remove them (I'll have a closer look when I return from vacation).</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>:</i>
<p><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel" class="user-hover" rel="iocanel">Ioannis Canellos</a> - what's the status of this?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=iocanel">iocanel</a>:</i>
<p>Apparently, I never had the chance to "give that closer look" and it has been pretty much forgotten.</p>
